### DISCUSSION TITLE
Where does System.err of Maven plugin go?

### DISCUSSION THREAD

**Discussion by basinilya** [eclipse-m2e/m2e-core#658] [Q&A] [answered]
Hi List.
I'm trying to debug the error (not reproducible with command line maven):
```
    JAXB errors arose while SchemaGen compiled sources to XML. (org.codehaus.mojo:jaxb2-maven-plugin:2.3.1:schemagen:schemagen:generate-resources)

    org.apache.maven.plugin.MojoExecutionException: JAXB errors arose while SchemaGen compiled sources to XML.
    	at org.codehaus.mojo.jaxb2.schemageneration.AbstractXsdGeneratorMojo.performExecution(AbstractXsdGeneratorMojo.java:385)
```
According to
https://github.com/mojohaus/jaxb2-maven-plugin/blob/a52f6c270f28753b4d92b71e36e629ad2c3a7ad8/src/main/java/org/codehaus/mojo/jaxb2/schemageneration/AbstractXsdGeneratorMojo.java#L411
it calls com.sun.tools.jxc.SchemaGenerator.run()

and according to the decompilation of SchemaGenerator$Runner it should print the errors to System.err before returning false:
```
/* 243 */       boolean res = task.call().booleanValue();
/*     */       
/* 245 */       for (Diagnostic<? extends JavaFileObject> d : diagnostics.getDiagnostics()) {
/* 246 */         System.err.println(d.toString());
/*     */       }
/* 248 */       return res;
```

I tried to run eclipsec.exe and I also checked the .log file in the Eclipse Workspace, but nothing is printed there. I also attached a debugger ti Eclipse, but the main Eclipse process doesn't load the above classes. Looks like m2e forks.

On the mailing list I was advised to install the M2E-Feature 'org.eclipse.m2e.logback.feature' which adds the Maven-Console that should show the print-out of Maven Plug-ins executed during the M2E-build within Eclipse.

I already had this feature installed, but the Maven Console did not show anything relevant.

Please tell me at least how to put a working breakpoint in SchemaGenerator$Runner. Is this really a forked process? How to start it in debug mode?

**Comment by mickaelistria** [comment]
Did you define some lifecycle mapping metadata for the execution of this mojo?  Mojo execution is usually *not* a subprocess.
To debug the Eclipse IDE, you need to install "Ecipse Plugin Development Environment" from the main site in Help > Install New Software. Then you can start a "child" Eclipse IDE which will be connected to Eclipse debugger, so you can try adding breakpoints to AbstractXsdGeneratorMojo .
HTH

**Comment by basinilya** [comment]
@mickaelistria  I'll try. Meanwhile do you know how to properly add maven plugin jars to debug classpath of a pde project? Should I manually add them to the classpath of the project named, say, "org.eclipse.m2e.maven.runtime"?

**Reply by mickaelistria** [reply]
I'm not sure what you're trying to achieve. The configuration of project is described in https://github.com/eclipse-m2e/m2e-core/blob/master/CONTRIBUTING.md#-developer-resources . I don't think you should try to modify classpath here as it's rules by OSGi/MANIFEST.MF, and we won't add into m2e support for random Maven plugins in the code `org.eclipse.m2e.maven.runtime` bundle. Only official or mainstream Maven plugins can be supported in core parts, other plugins will require to create a dedicated Eclipse plugin to support them.

**Reply by basinilya** [reply]
I mean org.codehaus.mojo.jaxb2-maven-plugin is not a dependency of any m2e plugins, but I need it in debug classpath to put a breakpoint in AbstractXsdGeneratorMojo when I choose Run -> Eclipse application

**Reply by mickaelistria** [reply]
I suggest you just import the code of jaxb2-maven-plugin in your workspace then. Make sure you pick the right tag!

**Comment by basinilya** [comment, marked as answer]
The problem is now solved.
* I didn't manage to reproduce it using Run As -> Eclipse application : doing so was adding some extra libraries to the eclipse classpath completely breaking JXC.
* Instead I added to `eclipse.ini, -vmargs`: `-Xrunjdwp:transport=dt_socket,address=0.0.0.0:7878,server=y,suspend=y`
and debugged the plugin project `org.eclipse.m2e.maven.runtime` as a remote java application
* I had to make sure that in the PDE workspace there're no installed JREs below the "justj" bundled with Eclipse for the JDK sources to be attached correctly
* For some reason the entry/exit breakpoints weren't hitting on `SchemaGenerator$Runner.compile()`, but they were on the task it was calling: `com.sun.tools.javac.api.JavacTaskImpl.call()`.  When the exit breakpoint hit I navigated to the parent frame and inpsected the local variable `diagnostics` and got the error messages I was looking for.

The reason why nothing was printed to System.err is that System.err is not used at all in the `SchemaGenerator` implemenation from `org.glassfish.jaxb:jaxb-jxc:2.2.11`. The local variable `diagnostics` is never read and its contents are lost.

So the solution is to upgrade to a more recent version of jaxb2-maven-plugin or jaxb-ri

The errors I was looking for were:

    import com.common.util.xml.HasLateValidate;
                              ^, C:\ftpjob\CommonJavaProject\src\com\common\ftpconfiguration\Subtask.java:13: error: package com.common.util.xml does not exist
    import com.common.util.xml.JAXB2ValidationEventCollector;
                              ^, C:\ftpjob\CommonJavaProject\src\com\common\ftpconfiguration\FtpConfig.java:23: error: package com.common.util.xml does not exist
    ...
